### DISCUSSION TITLE
Unknown type when type exists

### DISCUSSION THREAD

**Discussion by LorenzoTuri** [webonyx/graphql-php#841] [Q&A] [answered]
Hi everybody, I probably found an error with the arguments on nested types... 
Just to share some unnecessary info, what I'm trying to build is a CRUD entity agnostic controller with graphQL APIs in a symfony bundle. (The entities below (site, entitySiteUrl) are just entities taken from another project on wich I'm developing and testing)

I can't truly share the code since it's a little complicated to understand, but what I'm doing is to inject types, queries and mutations by creating them dynamically (php) into the schema, then quering and using resolve function to query doctrine.
On "normal" usage it works... for example this query
`query { entitySiteUrl (id: "01F2D393159DX6PFD7PAXD0YGH"){ id, url, site (baseUrl: "404") {id, baseUrl} } }`
Actually returns the resulting data (don't mind the fact the data is not actually filtered, that part needs to be build, but I need before the "nested filters" part to "build it better")\
result:
```json
{
  "data": {
    "entitySiteUrl": {
      "id": "01F2D393159DX6PFD7PAXD0YGH",
      "url": "urls",
      "site": {
        "id": "01F2D393159DX6PFD7PAXD0YGG",
        "baseUrl": "asdsadasdasdasdsad"
      }
    }
  }
}
```

The problem comes while building "complex" (maybe is better to say nested) filters...
`query { siteList ( siteUrls: { baseUrl: "123" } ){ id, baseUrl, siteUrls {id} } }`
```json
result:
{
  "errors": [
    {
      "message": "Unknown type: entitySiteUrl.",
      "extensions": {
        "category": "graphql"
      },
      "locations": [
        {
          "line": 1,
          "column": 9
        }
      ],
      "path": [
        "siteList"
      ],
      "trace": [
"other infos...."
      ]
}
```
This error only occurs with filtering the siteUrls fields (in this case baseUrl, but it's not different with other fields).
Also by using this query (the inverse side of the query before)
`query { entitySiteUrlList ( site: { url: "123" } ){ id, site {id} } }`
The code also returns a similar error (just with "site" instead of "entitySiteUrl")

Most of the schema creation code revolves simply around ObjectType and FieldArguments, so it should not incurr in strange lazy-loading mechanisms as far as I've seen in source code...

Il'' share the important part of the generated schema made from SchemaPrinter
```
scalar DateTime

type Mutation {
  siteUpdate(id: String!, baseUrl: String, dateCreate: DateTime, dateUpdate: DateTime, lastCheckDate: DateTime, checking: Boolean, siteUrls: [entitySiteUrl]): site
  siteCreate(id: String, baseUrl: String, dateCreate: DateTime, dateUpdate: DateTime, lastCheckDate: DateTime, checking: Boolean, siteUrls: [entitySiteUrl]): site
  siteDelete(id: String!, baseUrl: String, dateCreate: DateTime, dateUpdate: DateTime, lastCheckDate: DateTime, checking: Boolean, siteUrls: [entitySiteUrl]): site
  entitySiteUrlUpdate(id: String!, url: String, site: site): entitySiteUrl
  entitySiteUrlCreate(id: String, url: String, site: site): entitySiteUrl
  entitySiteUrlDelete(id: String!, url: String, site: site): entitySiteUrl
}

type Query {
  site(id: String!, baseUrl: String, dateCreate: DateTime, dateUpdate: DateTime, lastCheckDate: DateTime, checking: Boolean, siteUrls: [entitySiteUrl]): site
  siteList(id: String, baseUrl: String, dateCreate: DateTime, dateUpdate: DateTime, lastCheckDate: DateTime, checking: Boolean, siteUrls: [entitySiteUrl]): [site]
  entitySiteUrl(id: String!, url: String, site: site): entitySiteUrl
  entitySiteUrlList(id: String, url: String, site: site): [entitySiteUrl]
}

type entitySiteUrl {
  id: String
  url: String
  site(id: String, baseUrl: String, dateCreate: DateTime, dateUpdate: DateTime, lastCheckDate: DateTime, checking: Boolean, siteUrls: [entitySiteUrl], id: String, baseUrl: String, dateCreate: DateTime, dateUpdate: DateTime, lastCheckDate: DateTime, checking: Boolean, siteUrls: [entitySiteUrl]): site
}

type site {
  id: String
  baseUrl: String
  dateCreate: DateTime
  dateUpdate: DateTime
  lastCheckDate: DateTime
  checking: Boolean
  siteUrls: [entitySiteUrl]
}
```

In case you need other infos or whatever just tell me and I'll provide them

**Comment by spawnia** [comment, marked as answer]
1. Use PascalCase for types
2. Read up on input types vs output types: https://webonyx.github.io/graphql-php/type-definitions/

**Reply by LorenzoTuri** [reply]
Hi spawnia, thank you for the response!!

Actually changing the Case didn't result in any differences (in the results, the schema is obviusly now in PascalCase and the error is "Site" instead of "site" and "EntitySiteUrl" instead of "entitySiteUrl").

Also using InputObjectType to wrap ObjectType config when building FieldArgument::createMap also didn't solve anything, so I switched back to before...

Maybe it's best if I share the repository... just don't judge too much the code, it's still in developing phase ^^".
https://github.com/LorenzoTuri/SymfonyExtensions/tree/v1.0
The 2 incrimined files are:
// Controller class, calls the SchemaGenerator and pass the required parameters from request.
https://github.com/LorenzoTuri/SymfonyExtensions/blob/v1.0/GraphQLApi/Controller/GraphQLController.php
// Actually build the schema from a "list" of classes. -> the important part is generated aroud line 200. Arguments are injected later (because of some recursion issues) at line 120.
https://github.com/LorenzoTuri/SymfonyExtensions/blob/v1.0/GraphQLApi/Type/SchemaGenerator.php

I don't think the error has something to do with the code... Also loading the schema directly from the textual version (above) - that should be a valid schema as far as I know - Results in the same errors...

**Reply by spawnia** [reply]
The casing is just best practice.

I am saying you need to separate input and output types. You can not use a `type` in arguments or an `input` as field results.

**Reply by LorenzoTuri** [reply]
Oh, I've written it thinking in PHP cases 😅 okay thank you, I'll rewrite it in Pascal case...

Anyway the problem it still there 😅 also loading the schema by using the "textual version" didn't make results (using the textual version means the schema should be generated by this library using the proper input/output types, I suppose....)
I'll try debugging it more, I'll write results as soon as I have them

**Reply by LorenzoTuri** [reply]
Okay, I've found out what you meant, you were totally correct... 
I'll leave here some more info for others who may occur in this kind of error: 
This error may occur in case the developer didn't think/create the proper schema, and it's not related to errors in the library code, but in the way graphQL was made... for example, for my use case and entities the correct schema should be something like (pardon me for the camel case, still need to change it...):

```
"""
The `DateTime` scalar type represents time data, represented as an ISO-8601 (ATOM) encoded UTC date string.
"""
scalar DateTime

type Mutation {
  siteUpdate(id: String!, baseUrl: String, dateCreate: DateTime, dateUpdate: DateTime, lastCheckDate: DateTime, checking: Boolean, siteUrls: entitySiteUrlInput): site
  siteCreate(id: String, baseUrl: String, dateCreate: DateTime, dateUpdate: DateTime, lastCheckDate: DateTime, checking: Boolean, siteUrls: entitySiteUrlInput): site
  siteDelete(id: String!, baseUrl: String, dateCreate: DateTime, dateUpdate: DateTime, lastCheckDate: DateTime, checking: Boolean, siteUrls: entitySiteUrlInput): site
  entitySiteUrlUpdate(id: String!, url: String, site: siteInput): entitySiteUrl
  entitySiteUrlCreate(id: String, url: String, site: siteInput): entitySiteUrl
  entitySiteUrlDelete(id: String!, url: String, site: siteInput): entitySiteUrl
}

type Query {
  site(id: String!, baseUrl: String, dateCreate: DateTime, dateUpdate: DateTime, lastCheckDate: DateTime, checking: Boolean, siteUrls: entitySiteUrlInput): site
  siteList(id: String, baseUrl: String, dateCreate: DateTime, dateUpdate: DateTime, lastCheckDate: DateTime, checking: Boolean, siteUrls: entitySiteUrlInput): [site]
  entitySiteUrl(id: String!, url: String, site: siteInput): entitySiteUrl
  entitySiteUrlList(id: String, url: String, site: siteInput): [entitySiteUrl]
}

type entitySiteUrl {
  id: String
  url: String
  site(id: String, baseUrl: String, dateCreate: DateTime, dateUpdate: DateTime, lastCheckDate: DateTime, checking: Boolean, siteUrls: entitySiteUrlInput): site
}

input entitySiteUrlInput {
  id: String
  url: String
  site: site
}

type site {
  id: String
  baseUrl: String
  dateCreate: DateTime
  dateUpdate: DateTime
  lastCheckDate: DateTime
  checking: Boolean
  siteUrls: [entitySiteUrl]
}

input siteInput {
  id: String
  baseUrl: String
  dateCreate: DateTime
  dateUpdate: DateTime
  lastCheckDate: DateTime
  checking: Boolean
  siteUrls: [entitySiteUrl]
}
```

As you can see, and as spawnia wrote in the message before, the arguments and the fields should have a separated type. Even though in my use case "site" and "siteInput"  or "entitySiteUrl" and "entitySiteUrlInput" have the same type definition (same fields, same field types...), they have a different meaning for graphQL.
In the code, this is made by using InputObjectType instead of each ObjectType/ListOfType<ObjectType> in the fields.

Thank you again Spawnia =D =D
